rust/lance/src/io/exec/filtered_read.rs

```diff
@@ -0,0 +1,1584 @@
+            metrics,
+        })
+    }
+
+    fn obtain_stream(&self, partition: usize) -> SendableRecordBatchStream {
+        // There are two subtleties here:
+        //
+        // First, we need to defer execution until first polled (hence the once/flatten)
+        //
+        // Second, multiple partitions all share the same underlying task stream (see get_stream)
+        let running_stream_lock = self.running_stream.clone();
+        let dataset = self.dataset.clone();
+        let options = self.options.clone();
+        let metrics = self.metrics.clone();
+
+        let stream = futures::stream::once(async move {
+            let mut running_stream = running_stream_lock.lock().await;
+            if let Some(running_stream) = &*running_stream {
+                DataFusionResult::<SendableRecordBatchStream>::Ok(
+                    running_stream.get_stream(&metrics, partition),
+                )
+            } else {
+                let new_running_stream =
+                    FilteredReadStream::try_new(dataset, options, &metrics).await?;
+                let first_stream = new_running_stream.get_stream(&metrics, partition);
+                *running_stream = Some(new_running_stream);
+                DataFusionResult::Ok(first_stream)
+            }
+        })
+        .try_flatten();
+
+        Box::pin(RecordBatchStreamAdapter::new(self.schema(), stream))
+    }
+}
+
+impl DisplayAs for FilteredReadExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                let columns = self
+                    .options
+                    .projection
+                    .to_schema()
+                    .fields
+                    .iter()
+                    .map(|f| f.name.as_str())
+                    .collect::<Vec<_>>()
+                    .join(", ");
+                write!(
+                    f,
+                    "LanceRead: uri={}, projection=[{}], num_fragments={}, range_before={:?}, range_after={:?}, row_id={}, row_addr={}, indexed_filter={}, refine_filter={}",
+                    self.dataset.data_dir(),
+                    columns,
+                    self.options.fragments.as_ref().map(|f| f.len()).unwrap_or(self.dataset.fragments().len()),
+                    self.options.scan_range_before_filter,
+                    self.options.scan_range_after_filter,
+                    self.options.projection.with_row_id,
+                    self.options.projection.with_row_addr,
+                    self.options.filter_plan.index_query.as_ref().map(|i| i.to_string()).unwrap_or("true".to_string()),
+                    self.options.filter_plan.refine_expr.as_ref().map(|i| i.to_string()).unwrap_or("true".to_string()),
+                )
+            }
+        }
+    }
+}
+
+impl ExecutionPlan for FilteredReadExec {
```

Comment:
Did you mean to expose `metrics()` on here? I saw you hooked them up elsewhere.

In the future, would probably be good to implement:

* [repartitioned](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.repartitioned)
* [supports_limit_pushdown](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.supports_limit_pushdown), `with_fetch()`, and `fetch()`

Reply:
I've exposed them now in a future PR.  I haven't added support for `repartitioned` or `support_limits_pushdown` yet as I'll save that for future work.